### PR TITLE
feat: Adding Custom Url Endpoints and Headers

### PR DESCRIPTION
**Pull request by mstrehl**
## Related Issues/PRs

Adding custom base url and headers to access additional OpenAI endpoint deployments

## What changes are proposed in this pull request?
- Adding trait HasCustomHeader that checks if a customHeader was assigned.
- Adding customHeader to the addHeaders function to iterate through the Map of headers provided and assign them to the request.
- Added customRootUrl - if provided will be used instead of the prepareUrlRoot() function

## How is this patch tested?

- [X] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

## Does this PR change any dependencies?

- [ ] No. You can skip this section.
- [X] Yes. Make sure the dependencies are resolved correctly, and list changes here.
   - deploymentName parameter was changed to not required

## Does this PR add a new feature? If so, have you added samples on website?

- [X] No. You can skip this section.
- [ ] Yes. Make sure you have added samples following below steps.

1. Find the corresponding markdown file for your new feature in `website/docs/documentation` folder.
   Make sure you choose the correct class `estimators/transformers` and namespace.
2. Follow the pattern in markdown file and add another section for your new API, including pyspark, scala (and .NET potentially) samples.
3. Make sure the `DocTable` points to correct API link.
4. Navigate to website folder, and run `yarn run start` to make sure the website renders correctly.
5. Don't forget to add `<!--pytest-codeblocks:cont-->` before each python code blocks to enable auto-tests for python samples.
6. Make sure the `WebsiteSamplesTests` job pass in the pipeline.

### CODE REVIEW

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/services/CognitiveServiceBase.scala [hunk 1]
```diff
@@ -296,20 +322,25 @@ trait HasCognitiveServiceInput extends HasURL with HasSubscriptionKey with HasAA
...
+    getValueOpt(row, CustomHeader)
+  }
+
   protected def addHeaders(req: HttpRequestBase,
                            subscriptionKey: Option[String],
                            aadToken: Option[String],
                            contentType: String = "",
-                           customAuthHeader: Option[String] = None): Unit = {
+                           customAuthHeader: Option[String] = None,
+                           customHeader: Option[Map[String, String]] = None): Unit = {
```
Comment (mhamilton723) [inline cognitive/src/main/scala/com/microsoft/azure/synapse/ml/services/CognitiveServiceBase.scala:343]: nit: customHeaders

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/services/CognitiveServiceBase.scala [hunk 2]
```diff
@@ -188,6 +188,20 @@ trait HasCustomAuthHeader extends HasServiceParams {
   }
 }
 
+trait HasCustomHeader extends HasServiceParams {
+  // scalastyle:off field.name
+  val CustomHeader = new ServiceParam[Map[String, String]](
```
[the remark this answers is not in the document]
Comment (mstrehl) [reply cognitive/src/main/scala/com/microsoft/azure/synapse/ml/services/CognitiveServiceBase.scala]: fixed

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/services/CognitiveServiceBase.scala (hunk 1 above)
Comment (mstrehl) [reply cognitive/src/main/scala/com/microsoft/azure/synapse/ml/services/CognitiveServiceBase.scala]: fixed

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/services/CognitiveServiceBase.scala [hunk 3]
```diff
@@ -188,6 +188,20 @@ trait HasCustomAuthHeader extends HasServiceParams {
   }
 }
 
+trait HasCustomHeaders extends HasServiceParams {
+  // scalastyle:off field.name
+  val CustomHeaders = new ServiceParam[Map[String, String]](
+    this, "CustomHeader", "List of Custom Header Key-Value Tuples."
```
Comment (mhamilton723) [inline cognitive/src/main/scala/com/microsoft/azure/synapse/ml/services/CognitiveServiceBase.scala:194]:
```suggestion
    this, "customHeaders", "Dictionary of Custom Header Key-Value Tuples."
```

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/services/CognitiveServiceBase.scala [hunk 4]
```diff
@@ -188,6 +188,20 @@ trait HasCustomAuthHeader extends HasServiceParams {
...
+  )
+  // scalastyle:on field.name
+
+  def setCustomHeaders(v: Map[String, String]): this.type = {
+    setScalarParam(CustomHeaders, v)
+  }
+
+  def getCustomHeader: Map[String, String] = getScalarParam(CustomHeaders)
+}
+
```
Comment (mhamilton723) [inline cognitive/src/main/scala/com/microsoft/azure/synapse/ml/services/CognitiveServiceBase.scala:204]:
```suggestion
trait HasCustomHeaders extends HasServiceParams {

    val customHeaders = new ServiceParam[Map[String, String]](
    this, "customHeaders", "Map of Custom Header Key-Value Tuples."
  )

  def setCustomHeaders(v: Map[String, String]): this.type = {
    setScalarParam(customHeaders, v)
  }

  def getCustomHeaders: Map[String, String] = getScalarParam(customHeaders)
}

```

File: cognitive/src/test/scala/com/microsoft/azure/synapse/ml/services/openai/OpenAIChatCompletionSuite.scala [hunk 5]
```diff
@@ -151,6 +151,29 @@ class OpenAIChatCompletionSuite extends TransformerFuzzing[OpenAIChatCompletion]
     assert(Option(results.apply(2).getAs[Row]("out")).isEmpty)
   }
 
+  test("Custom EndPoint") {
```
Comment (mhamilton723) [inline cognitive/src/test/scala/com/microsoft/azure/synapse/ml/services/openai/OpenAIChatCompletionSuite.scala:154]:
```suggestion
  ignore("Custom EndPoint") {
```

**Review by mhamilton723 — CHANGES_REQUESTED** [review]
Put a few changes in for you other than that if everything passes after should be good to go

**Pull request merged**